core/src/org/pentaho/di/core/encryption/CertificateGenEncryptUtil.java

```diff
@@ -53,10 +62,52 @@ public static KeyPair generateKeyPair() {
     return pair;
   }
 
+  public static Key generateSingleKey() throws NoSuchAlgorithmException {
+    Key key = KeyGenerator.getInstance( SINGLE_KEY_ALGORYTHM ).generateKey();
+    return key;
+  }
+
+  public static byte[] encodeKeyForTransmission( Key encodingKey, Key keyToEncode ) throws Exception {
+    Cipher cipher = Cipher.getInstance( TRANSMISSION_CIPHER_PARAMS );
+    cipher.init( Cipher.WRAP_MODE, encodingKey );
+    byte[] encodedKey = cipher.wrap( keyToEncode );
+    return encodedKey;
+  }
+
+  public static Key decodeTransmittedKey( byte[] sessionKey, byte[] transmittedKey, boolean privateKey )
+      throws Exception {
+    KeySpec keySpec = null;
+    Key keyKey = null;
+    if ( transmittedKey == null || sessionKey == null ) {
+      return null;
+    }
+    if ( !privateKey ) {
+      keySpec = new X509EncodedKeySpec( sessionKey );
+      keyKey = KeyFactory.getInstance( PUBLIC_KEY_ALGORYTHM ).generatePublic( keySpec );
+    } else {
+      keySpec = new PKCS8EncodedKeySpec( sessionKey );
+      keyKey = KeyFactory.getInstance( PUBLIC_KEY_ALGORYTHM ).generatePrivate( keySpec );
+    }
+    Cipher keyCipher = Cipher.getInstance( TRANSMISSION_CIPHER_PARAMS );
+    keyCipher.init( Cipher.UNWRAP_MODE, keyKey );
+    return keyCipher.unwrap( transmittedKey, SINGLE_KEY_ALGORYTHM, Cipher.SECRET_KEY );
+  }
+
+  public static Cipher initDecryptionCipher( Key unwrappedKey, byte[] unencryptedKey ) throws Exception {
+    Cipher decryptionCip = Cipher.getInstance( SINGLE_KEY_ALGORYTHM );
+    if ( unwrappedKey != null ) {
+      decryptionCip.init( Cipher.ENCRYPT_MODE, unwrappedKey );
+    } else {
+      SecretKeySpec sks = new SecretKeySpec( unencryptedKey, SINGLE_KEY_ALGORYTHM );
+      decryptionCip.init( Cipher.ENCRYPT_MODE, sks );
+    }
+    return decryptionCip;
+  }
+
   public static byte[] encryptUsingKey( byte[] data, Key key ) {

```

Comment:
We should catch the most likely exceptions and craft some readable error messages for them rather than just relying on call-stacks.  These errors are likely to confuse users and we don't want to frustrate them.